### ISSUE TITLE
Starcounter as a service?

### ISSUE THREAD

**Issue by careri** [Starcounter/Starcounter#3852]
Is it possible to install starcounter as a service?
I think it used to be an option in the installer, but it's no longer there?

**Comment by per-samuelsson** [comment]
@careri 

Quite some time since this was discussed the last time around, but AFAIK, there is no principal problem why it shouldn't be possible to do that.

Here is a relevant https://github.com/Starcounter/Starcounter/issues/2928#issuecomment-145775228 on that topic.

Just be sure you understand debugging the code host (and applications inside it) will be a pain when run in that fashion.

As for the installer, I need to ping @alemoi.

**Comment by alemoi** [comment]
There's a practical case when @ubbek installed it as a service, as I remember. @per-samuelsson there are certain conditions when its installed as a service, that wasn't changed. Just need to look for tje exact conditions, it was related to VS plugin, as I remember.

**Comment by ubbeK** [comment]
@alemoi Yes, I had to start the `scservice.exe` as a service when I ran a Jenkins Continuous Integration server as a service.

What I did was to create a service through a console with the command (note that sc.exe is the windows Service Controller)
`sc.exe create MyStarcounterService binPath="C:\Program Files\Starcounter\scservice.exe"`

After that one may simply start `scservice.exe` as a service by the following command
`sc.exe start MyStarcounterService`

From there on I could write any normal Starcounter commands on the CI server, for example
`star.exe MyStarcounterApp.exe`

**Comment by alemoi** [comment]
I have reviewed the code and Starcounter always installs as a service, which has manual start if VS extensions are also installed, because of the following reason:

> Just be sure you understand debugging the code host (and applications inside it) will be a pain when run in that fashion.

@careri can you please clarify, maybe your installation does not create the service like this?

![image](https://cloud.githubusercontent.com/assets/208134/18769180/31c1ccb0-812b-11e6-940d-1620b3c0ea97.png)

**Comment by careri** [comment]
@ubbeK Just curious, why would you run starcounter on the builder server (Jenkins)?
I made some custom NuGets containing the necessery DLLs to build starcounter projects. If no StarcounterBin Environment variable is defined I redirect the assembly referenes to the ones in my NuGet
> Yes, I had to start the scservice.exe as a service when I ran a Jenkins Continuous Integration server as a service.

**Comment by careri** [comment]
@alemoi I think I might have missed something here. 
scservice.exe - The monitor/manager of other starcounter processes
I have two instances of this process. One running as system and one running as me.

When is the one running as system used? The one bound to the registered windows service that is.

I can stop that service and still use star.exe, start/stop from Visual Studio.


The reason for me asking is, how would we keep our starcounter app running at a customer.
Should I write my own service that calls star.exe with my application as argument?

**Comment by per-samuelsson** [comment]
> I have two instances of this process. One running as system and one running as me.

Really? I thought that was not even possible. It's certainly not desired, if I haven't missed on something new.

**Comment by per-samuelsson** [comment]
> The reason for me asking is, how would we keep our starcounter app running at a customer.
Should I write my own service that calls star.exe with my application as argument?

Can you elaborate some more on this? I don't understand it.

As we modelled it initially, the idea was that in production environments, Starcounter would bootstrap as a Windows service. And as a consequence of that, code hosts would be run in the same context, i.e. in Windows session 0. Do you see a problem with that model? If so, what?

**Comment by per-samuelsson** [comment]
> I have reviewed the code and Starcounter always installs as a service, which has manual start if VS extensions are also installed, because of the following reason:

Exactly. That is the design we choose. 

We actually discussed surfacing this in the installer - like have a radio-box combo where you would choose either "Production" or "Development" installation. If you choose "Production", we would configure the Windows service running Starcounter to **auto start**. If you choose "Development", we would instead set the service to **manual start**. And the later would result in that the first invocation of our tools (e.g. `star myapp.exe`) would start the `scservice.exe` on demand, running in user-mode (i.e. Session > 1).

Instead, we choose to keep the same logic, but to determine if the installation was a production- or development-installation based on if any Visual Studio extension was installed or not. Rationale: if you install that extension, you are probably not on a production server.

**Comment by careri** [comment]
We'll might be that there is some info I haven't read. But please explain to me how we would go about running in production?

This is the Install procedure for production:
1. Install stable 2.2
2. Extract our code into a directory
3. Create a database in the Administrator (called HAD in this case)
4. Register/start our application using star.exe and the database HAD


Seems correct so far? But now what?
Say that the computer restarts, the starcounter service will pull upp the scservice.exe. But will it also start the HAD database and the registered application?

**Comment by careri** [comment]
> Really? I thought that was not even possible. It's certainly not desired, if I haven't missed on something new.

1. Start the Starcounter service it will spawn a scservice in system session.
2. "Starcounter Personal Service", right-click run as administrator
![image](https://cloud.githubusercontent.com/assets/4263597/18776505/edb1428c-8169-11e6-8292-593dc7ebd4e9.png)

**Comment by per-samuelsson** [comment]
> Seems correct so far?

Yep.

> Say that the computer restarts, the starcounter service will pull upp the scservice.exe. But will it also start the HAD database and the registered application?

No. We don't have restart functionality like that built in to my knowledge. The (admin) server will start (because it's configured as an auto-starting service), but you will need to start your databases / applications as in [4] from the above list.


> Shall I never use the "Starcounter Personal Server" on the start menu [...] 

Normally not, no.

> [...] maybe it's not even available in production?

Not sure actually. Maybe @alemoi knows. No matter what, just ignore it if so.

**Comment by per-samuelsson** [comment]
> 1. Start the Starcounter service it will spawn a scservice in system session.
> 2. "Starcounter Personal Service", right-click run as administrator 

I'm surprised this didn't result in that the personal server refused to start (because an instance of `scservice` and related backend-processes already execute, and ports would be occupied. But again, I haven't been part of this area for quite some time, so maybe something has changed and this is a supported state.

**Comment by per-samuelsson** [comment]
> Create a database in the Administrator (called HAD in this case)

FYI, you can create databases in the CLI as well: `staradmin new db`. Just in case you find a need to.

**Comment by careri** [comment]
> No. We don't have restart functionality like that built in to my knowledge. The (admin) server will start (because it's configured as an auto-starting service), but you will need to start your databases / applications as in [4] from the above list.

That's why I'm suggesting writing my own service that is depending on the starcounter service. In the start-callback I would launch star.exe with the arguments needed for my application.

And in Stop I would just stop my database HAD

**Comment by careri** [comment]
Could you please explain why debugging processes running in System session would be a problem?
As long as I run VS as administrator I can see no problem?

**Comment by ubbeK** [comment]
> @ubbeK Just curious, why would you run starcounter on the builder server (Jenkins)?
> I made some custom NuGets containing the necessery DLLs to build starcounter projects. If no StarcounterBin Environment variable is defined I redirect the assembly referenes to the ones in my NuGet

@per-samuelsson 
Sure! Because we use the same server as a test server and we wanted to use the same scipts for build and deploy.
NuGet, what version of Starcounter is there? Latest Stable?

**Comment by per-samuelsson** [comment]
> Could you please explain why debugging processes running in System session would be a problem?
> As long as I run VS as administrator I can see no problem?

Yeah, that should work. I just remember we had problems with different privileges and how to enable debugging in Windows in the past, when trying to debug Session 0 processes. Hence the different two runtime modes. Feel free to do whatever you find necessary, it's just not under our "super-easy" umbrella. 🌂

**Comment by careri** [comment]
@ubbeK I just made a NuGet for Heads (my company) for the needed assemblies.
Were currently at 2.2 stable. But moving to the new stable as soon as it's released.

So I'll make a new one then.

**Comment by alemoi** [comment]
> Shall I never use the "Starcounter Personal Server" on the start menu, maybe it's not even available in production?

This is only for manual start, if needed by some reason.

> Seems correct so far? But now what?
Say that the computer restarts, the starcounter service will pull upp the scservice.exe. But will it also start the HAD database and the registered application?

As far as I know, @andwah maybe has some info about this. I thought it was implemented in one of our installations, but not sure.

**Comment by andwah** [comment]
> Seems correct so far? But now what?
> Say that the computer restarts, the starcounter service will pull upp the scservice.exe. But will it also start the HAD database and the registered application?

There is no auto start of databases when scservice starts.

**Comment by alemoi** [comment]
> There is no auto start of databases when scservice starts.

Ok, I guess I confused it with auto-start of apps, when starting the database, right?

**Comment by careri** [comment]
Ok, sounds like I'll implement my own service then.
You can close this issue if you don't have any internal questions based on this

**Comment by careri** [comment]
> Ok, I guess I confused it with auto-start of apps, when starting the database, right?

How does the auto-start of apps feature work?

**Comment by alemoi** [comment]
> How does the auto-start of apps feature work?

@andwah I am not sure if it was just experimental. But you can elaborate on this better.

**Comment by andwah** [comment]
> How does the auto-start of apps feature work?

When a database is started from the admin gui all apps that is "auto-started" will also start.

Note: The apps will not be started if the database is started with star.exe or visual studio (f5).

**Comment by careri** [comment]
So to get them autostarted you need to register them through the Web UI? Or can this be achieved from the commandline?
Production wise it would be nice if an installer could do all of this without using the Web UI.

**Comment by andwah** [comment]
@careri You can use star.exe to start your apps from a batch file.

**Comment by careri** [comment]
@andwah Yes that is what I'm doing now. 
But I was asking how this autostart feature works, is it only if you register through the Web UI?

I'll probably end up writing a service that starts stops the database/app

**Comment by andwah** [comment]
@careri yes

> When a database is started from the admin gui all apps that is "auto-started" will also start.
> 
> Note: The apps will not be started if the database is started with star.exe or visual studio (f5).

`Auto-started` is a button on each application in the admin gui aka Web UI.

There is an undocumented REST API to Start a database and it's 'auto-started' applications. And we also have a dll (in a private repo at the moment) for communicating with the server

**Comment by careri** [comment]
I'm guessing this Auto-started is a new feature not in 2.2-stable? Because I can't find it anywhere

**Comment by andwah** [comment]
Cant remember when it was introduced.

Here is a screenshot of my admin gui (latest starcounter version).

![image](https://cloud.githubusercontent.com/assets/2656996/18785734/3b9fc1f0-819a-11e6-9fdf-31e3e06aee50.png)

**Comment by careri** [comment]
Ah, that explains it. My App isn't that kind of App. It's just an executable, not a micro app.

**Comment by andwah** [comment]
@careri ahh :-) .  Then with just an executable your right. they can not be auto-started. 
The scservice/database does not know that they exist until they are started.

So i recommend to use `star.exe` to start/stop the database and applications.

**Comment by Mnosek** [comment]
> @careri ahh :-) . Then with just an executable your right. they can not be auto-started. 
The scservice/database does not know that they exist until they are started.

I've just faced the same problem, that I can't auto-start apps which are not installed from the warehouse (executables). On a test environment I need to write scripts manually to start everything. It would be very useful to mark apps executed once as auto-starting. Maybe we should consider implementing it in the future?

**Comment by andwah** [comment]
@Mnosek It's possible already today to add an local executable to be autostarted (via a configuration file).  

It's **not user friendly** as you need to edit the configuration file manually.

**AND** you need to start the database from the administrator gui (or with a REST call). 

I can help you to set this up if this is the way you like to go.

**Comment by Mnosek** [comment]
Hi @andwah, thanks for your reply. 
My goal is to run Starcounter and executables automatically if the process is shut down or machine is restarted. For now I do it by custom watchdog running as a service and scheduled system task to run everything after startup. Batch file is responsible for starting Starcounter server and executables. Unfortunately it isn't a good solution, there is no way to make this configuration work if Windows Data Execution Prevention security mechanism is working.

My point is that we don't necessarily need to have a nice gui. It is Admin's work, so it might be manual configuration file. It would be great if you can help me to set it up!

**Comment by dan31** [comment]
@andwah Do you think you could wrap up the use of manually autostarted executables as an .io article? All our users would largely appreciate that! I would suggest this section: http://starcounter.io/guides/tools/.

@Mnosek which watchdog do you use? There is one https://github.com/dan31/cbwatchdog that I've made to substitute the absense of a watchdog process in Starcounter 2.x. This is a temporary solution which works. I hope we'd address it properly in upcoming releases (starting from 2.5 I believe). I think the switch to ```bool elevatedModeRecovery = true;``` should do the trick for Windows DEP, shoudn't it?

**Comment by andwah** [comment]
@Mnosek The configuration file `applications.json` is located in the database repository e.i. `\Documents\Starcounter\Personal\Databases\mydatabase\applications.json`. If there is no autostarted apps the file will not exist. And you need to created it manually.

**applications.json**  example:
```
{
	"Deployed": [{
		"Namespace": "StarcounterSamples.SignIn",
		"AppName": "SignIn",
		"Channel": "Stable",
		"Version": "5.0.1",
		"Arguments": ""
	}],
	"Local": [{
		"AppName": "StarcounterApplication1",
		"Executable": "c:\\code\\Starcounter\\StarcounterApplication1\\bin\\Debug\\StarcounterApplication1.exe",
		"Arguments": "",
		"Resourcefolder": "c:\\code\\Starcounter\\StarcounterApplication1\\wwwroot\\"
	}]
}
```

The ***Deployed*** array is populated by the administrator gui.
The ***Local*** array is where you manually add your own executable to be auto started when the database is started from the Administrator gui or with a REST call.

How to start a database with REST

***StartDatabase***
Post a `startdatabase` task request and poll the task until status is 0.

Method: POST
URL: /api/tasks/startdatabase
Port: 8181 (Default administrator port)
Content (startdatabase.json):
```
{
	"DatabaseName" : "mydatabase"
}
```

Response:
Statuscode: 201 (Created)

Content: (Task.json)
```
{
  "ID": "",  // Task id
  "Status": 0,   // 0 OK,  < 0 Error, check Message
  "Message": ""
   ...
}

```

***Get Task***

Method: GET
URL /api/tasks/{ID}
Content: (Task.json)

Note: i have not tested this REST as im writing this from my memory :/ i will check it better when i get back to office on monday.

**Comment by Mnosek** [comment]
@dan31 I've used exactly this watchdog made by you :)
>  I hope we'd address it properly in upcoming releases (starting from 2.5 I believe)

Perfect!


>  I think the switch to bool elevatedModeRecovery = true; should do the trick for Windows DEP, shoudn't it?

Unfortunately it does not. I've checked once again. When DEP is turned on, Starcounter server starts, but apps (executables) doesn't. Service is set up for "Automatic (Delayed)" start, log on as "Local System Account". **Actually the problem occurs only when I restart the machine**. If I just kill all Starcounter processes by `staradmin kill all` everything (server and executables) starts correctly. Do we have any working example, when I can check the configuration? Maybe I'm missing something like user, priviliges?

**Comment by andwah** [comment]
> Do you think you could wrap up the use of manually autostarted executables as an .io article?

@dan31 Im not sure. it's very internal use and if we are going to make it public then i think we need to solve on how to start a database first.

**Comment by Mnosek** [comment]
@andwah thank you, I will take a look at it and try it out .
